### ISSUE TITLE
Performance of large word lists

### ISSUE THREAD

**Issue by olbrich** [gchan/password_blacklist#3]
I ended up creating a similar tool for one of my projects that uses the larger and more comprehensive word lists found here: https://github.com/berzerk0/Probable-Wordlists.

Dealing with such large lists is problematic because of their sheer size as this will require loading a large number of strings into memory or doing an operating system level comparison on the file.  I found that I could get reasonable performance by recognizing that a large number of the words in the large lists were invalid due to other restrictions (i.e., not long enough, no number, no symbol, etc...).

**Comment by gchan** [comment]
Thanks for the tip, the project you suggested seems to have a more comprehensive word list! I will definitely take a look at replacing my current wordlist with one from over there.

Good idea pruning the wordlists based on your password restrictions (length and character type requirements). Unfortunately for this project, I can't assume there are any restrictions 😬 (perhaps I can assume there is a length restriction). So at the cost of additional time/space, I have decided to make this library as general as possible without any assumption of password restrictions.

In additional to any password restrictions, it also might be nice to let the user know they are using a common known password. This choice to left to the implementer.

**Comment by olbrich** [comment]
Another suggestion for you... instead of trying to use a regex to search the list as a giant string, you might do the following...

1. load in the list as a sorted array
2. use `bsearch` to look for the potentially bad password.

 ```
common_passwords = File.readlines(filename).map(&:chomp).map(&:downcase).uniq.sort 
# the sort, downcase and unique could be done on the file ahead of time for efficiency

password_found = !common_passwords.besearch { |common_password| password.downcase <=> common_password }.nil?
```

This will do a binary search through the entire list, which should be much faster than a regex match. (my very basic experiments suggest it's about 5 times faster).

I would also do the downcase so that you are doing case insensitive matching (`PaSsWoRd` isn't really better than `password`).

**Comment by gchan** [comment]
Thanks for the suggestion, I had issues using `Array#bsearch` but I found using using Ruby `Set` to be extremely quick. There is probably a trade-off with additional space (memory) but worth investigating. 

PRs are welcome :) There is a benchmark file to compare the various ways to detect a bad password if you want to compare your solution https://github.com/gchan/password_blacklist/blob/master/bin/benchmark

**Comment by olbrich** [comment]
Yeah, sets are even more efficient.  Use that.

**Comment by gchan** [comment]
Some benchmark results:
```
Comparison:
        Set#include?:  4847468.9 i/s
       Array#bsearch:   497979.7 i/s - 9.73x  slower
        String#match:     5164.2 i/s - 938.67x  slower
```

The downside is that a `Set` requires a significant amount of memory (I believe it was more than 10x) so I decided to go for the more space efficient solution for now. However there is no reason why we couldn't write a fast version that uses `Set` (e.g. `PasswordBlacklist::FastChecker`).

I've taken your suggestion of performing case insensitive matching. This will be released in version 0.2.0.
